File: tests/fixtures/quality/scala/Scheduler.scala
package com.acme.worker

/** Anything the scheduler can drive. */
trait Runnable {

  /** Begin processing. */
  def start(): Unit

  def depth: Int
}

/** A unit of deferred work. */
case class JobSpec(name: String, attempts: Int)

/** Runs jobs with exponential backoff between retries. */
class Scheduler(val queue: String) extends BaseWorker with Runnable {

  /** Bounds how often a job is retried. */
  val maxAttempts: Int = 5

  private var pending: List[JobSpec] = Nil

  /** Drain the pending queue. */
  def start(): Unit = pending.foreach(spec => dispatch(spec, maxAttempts))

  /** Run one job, retrying with exponential backoff. */
  def dispatch(
      spec: JobSpec,
      max: Int
  ): Boolean = {
    var attempt = 0
    while (attempt < max) attempt += 1
    true
  }

  private def reset(): Unit = pending = Nil

  def depth: Int = pending.size
}

/** Factory helpers. */
object SchedulerFactory {

  /** Build an empty scheduler. */
  def build(queue: String): Scheduler = new Scheduler(queue)
}

File: tests/fixtures/quality/ruby/scheduler.rb
# Runs queued jobs with a retry budget.
module Worker
  # Bounds how often a job is retried.
  MAX_ATTEMPTS = 5

  # Anything the scheduler can drive.
  module Runnable
    # Begin processing.
    def start
      raise NotImplementedError
    end
  end

  # One queued unit of work.
  JobSpec = Struct.new(:name, :attempts) do
    # Whether the budget allows another attempt.
    def retry?
      attempts < MAX_ATTEMPTS
    end
  end

  # Runs jobs with exponential backoff between retries.
  class Scheduler < BaseWorker
    include Runnable

    attr_reader :queue

    # Build an empty scheduler.
    def initialize(queue)
      @queue = queue
      @pending = []
    end

    # Drain the pending queue.
    def start
      @pending.each { |spec| dispatch(spec) }
    end

    # Run one job, retrying with exponential backoff.
    def dispatch(spec, max_attempts = MAX_ATTEMPTS)
      attempt = 0
      attempt += 1 while attempt < max_attempts
      true
    end

    def self.build(queue)
      new(queue)
    end

    class << self
      # The scheduler shared by every caller that does not bring its own.
      def default
        @default ||= build("default")
      end
    end

    # Heavier jobs wait longer between attempts.
    protected def weight(spec)
      spec.size
    end

    private

    def reset
      @pending.clear
    end
  end
end
